### PR TITLE
refactor: split widget state into several hooks and added new lifecycle hook

### PR DESCRIPTION
**Pull request by WANZARGEN**
### To Reviewers
- [ ] Skip (`style`, `chore`, `ci` ONLY)
- [ ] Not that difficult

### Type of Change
- [ ] New feature
- [ ] Bug fixes
- [ ] Feature improvement
- [x] Refactor
- [ ] Others (performance improvement, CI/CD, etc.)

### Affects to
- [ ] Packages
  - [ ] core-lib
  - [ ] mirinae
  - [ ] etc 

- [ ] Apps
  - [ ] storybook
  - [x] web

### Checklist
- Did you check the `lint` and `type`?
- Did you document the changes?
  - Changes in `mirinae` should be reflected in `storybook`.
- Did you test the app after the package changes?

### Description

- Split widget state hook into several hooks
- Added useWidget to integrate hooks that always used together (frame props, widget state)
- Applied to CostMapWidget

### CODE REVIEW

File: apps/web/src/services/dashboards/shared/dashboard-widget-container/DashboardWidgetContainer.vue [hunk 1]
```diff
@@ -22,7 +23,33 @@
...
+                           :edit-mode="editMode"
+                           :error-mode="editMode && dashboardDetailState.widgetValidMap[widget.widget_key] === false"
+                           :all-reference-type-info="allReferenceTypeInfo"
+                           :initiated="!!initiatedWidgetMap[widget.widget_key]"
+                           :disable-refresh-on-variable-change="dashboardDetailState.widgetViewModeModalVisible && widget.widget_key !== widgetFormState.widgetKey"
+                           @update-data="handleUpdateData(widget.widget_key, $event)"
+                           @update-widget-info="handleUpdateWidgetInfo(widget.widget_key, $event)"
+                           @update-widget-validation="handleUpdateValidation(widget.widget_key, $event)"
+                />
+                -->
```
Comment (yuda110) [inline apps/web/src/services/dashboards/shared/dashboard-widget-container/DashboardWidgetContainer.vue:52]: question: Will this be updated next?

File: apps/web/src/services/dashboards/shared/dashboard-widget-container/DashboardWidgetContainer.vue [hunk 2]
```diff
@@ -22,7 +23,33 @@
                            :error-mode="editMode && dashboardDetailState.widgetValidMap[widget.widget_key] === false"
                            :all-reference-type-info="allReferenceTypeInfo"
                            :initiated="!!initiatedWidgetMap[widget.widget_key]"
+                           :disable-refresh-on-variable-change="dashboardDetailState.widgetViewModeModalVisible"
+                           :dashboard-settings="dashboardDetailState.settings"
+                           :dashboard-variables-schema="dashboardDetailState.variablesSchema"
+                           :dashboard-variables="dashboardDetailState.variables"
+                           @update-data="handleUpdateData(widget.widget_key, $event)"
+                           @update-widget-info="handleUpdateWidgetInfo(widget.widget_key, $event)"
+                           @update-widget-validation="handleUpdateValidation(widget.widget_key, $event)"
```
Comment (yuda110) [inline apps/web/src/services/dashboards/shared/dashboard-widget-container/DashboardWidgetContainer.vue:32]: praise: 👍

File: apps/web/src/services/dashboards/widgets/_hooks/use-widget/use-cost-widget.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+ const { widgetState, widgetFrameProps } = useCostWidget(props, {
+    widgetLocation: computed<Location>(() => ({
+        name: COST_EXPLORER_ROUTE.COST_ANALYSIS._NAME,
+    })),
+    dateRange: computed<DateRange>(() => {
+        const end = widgetState.settings?.date_range?.end ?? dayjs.utc().format('YYYY-MM');
+        const start = widgetState.settings?.date_range?.start ?? dayjs.utc().format('YYYY-MM');
+        return { start, end };
+    }),
+ });
```
Comment (yuda110) [inline apps/web/src/services/dashboards/widgets/_hooks/use-widget/use-cost-widget.ts:30]: praise: 👍

**Review by yuda110 — APPROVED** [review]
LGTM. Thank you for your refactoring!!!!!!

**Review by piggggggggy — APPROVED** [review]
LGTM! 😧😧
There are some parts difficult for me. Could you share it with us at next sharing time?

File: apps/web/src/services/dashboards/shared/dashboard-widget-container/DashboardWidgetContainer.vue (hunk 1 above)
Comment (WANZARGEN) [reply apps/web/src/services/dashboards/shared/dashboard-widget-container/DashboardWidgetContainer.vue:52]: yes!

**Comment by WANZARGEN** [comment]
> LGTM! 😧😧 There are some parts difficult for me. Could you share it with us at next sharing time?

Of course!

**Pull request merged**